Add TacoClient + Configuration Validation (#700)
nucypher/taco-web: packages/taco/README.md

For detailed viem documentation, see [VIEM_SUPPORT.md](./VIEM_SUPPORT.md).

## TacoClient - Object-Oriented Interface

For applications requiring multiple TACo operations or complex configuration management, the TACo SDK provides an optional object-oriented interface through the `TacoClient` class. This provides a stateful, higher-level abstraction over the functional API.

### What is different from the functional API?

- **Stateful Configuration**: Store domain, ritual ID, and other client configurations at instance level so there is less need to repeatedly pass some config parameters.
- **Enhanced Validation**: Enhanced configuration validation and correction.

> - **Enhanced Type Safety**: More type strict for the domain name.

Reviewer: Consider to remove these explanations. IMHO what a developer expects from a README is a quick start guide, and not a justification about why we decided to implement a feature. This feels like noise to me.

Personal opinion, of course it's up to you.

```suggestion
```

The Object-Oriented API is fully backward compatible - you can use both APIs in

> the same application as needed. Except that the TacoClient has more validations

Reviewer 2: ```suggestion
the same application as needed. Except that the TacoClient has additional validations
```

and hence throws some errors earlier with different error messages.

> NOTE: Using TacoClient is basically similar to using the functional API. And there is no

Reviewer 2: ```suggestion
NOTE: Using TacoClient is equivalent to using the functional API. And there is no
```

specific recommendations on what to use. It is for the convenience of the

> developer to choose.

Reviewer 3: Additional iteration:
```suggestion
NOTE: Using `TacoClient` is equivalent to using the functional API. 
There are no specific recommendations on which approach to use. 
Choose the one that best suits your development preferences.
```

### Basic Usage

```typescript
import { TacoClient, ConditionContext, DOMAIN_NAMES } from '@nucypher/taco';
import { createPublicClient, http } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { polygonAmoy } from 'viem/chains';

// Initialize TACo
await initialize();

// Set up viem client and account
const viemClient = createPublicClient({
  chain: polygonAmoy,
  transport: http(),
});
const viemAccount = privateKeyToAccount('0x...');

// Create TacoClient instance with domain constants
const tacoClient = new TacoClient({

>   domain: DOMAIN_NAMES.TESTNET, // TacoDomains.TESTNET -> 'tapir'

Reviewer 3: ```suggestion
  domain: DOMAIN_NAMES.TESTNET, // TESTNET -> 'tapir'
```
`TacoDomains` is undefined. I guess you just wanted to indicate that our testnet is "tapir"?

Author: Sorry, this was an old name that I changed but missed here. Thanks for the catch.

  ritualId: 6,
  viemClient,
  viemAccount
});

// Encrypt data
const messageKit = await tacoClient.encrypt('Hello, secret!', condition);

// Decrypt
const conditionContext = ConditionContext.fromMessageKit(messageKit);

// if needed Add authentication for ":userAddress" in condition...

const decryptedMessage = await tacoClient.decrypt(messageKit, conditionContext);
// OR with encrypted bytes:
// const decryptedMessage = await tacoClient.decrypt(messageKit.toBytes(), conditionContext);
```

### Dual Configuration Support

TacoClient supports both viem and ethers.js configurations:

```typescript
import { TacoClient, DOMAIN_NAMES } from '@nucypher/taco';

// With viem
const tacoClientViem = new TacoClient({
  domain: DOMAIN_NAMES.TESTNET,
  ritualId: 6,
  viemClient,
  viemAccount
});

// With ethers.js
const tacoClientEthers = new TacoClient({
  domain: DOMAIN_NAMES.TESTNET,
  ritualId: 6,
  ethersProvider,
  ethersSigner
});
```

> ### Backward Compatibility
> 
> The TacoClient provides a higher-level interface while maintaining full backward compatibility:
> 
> - **Zero Breaking Changes**: Existing functional API remains unchanged
> - **Optional**: Using TacoClient is optional. The functional API is still there.
> - **Interoperable**: Can use both APIs in the same application, if needed.
> 

Reviewer 2: This feels like a repeat to me.
```suggestion
```

Reviewer 3: Agreed.
